Feature : New Trees Placement Algorithm "Forest".

SylvainDevidal opened the pull request:
It creates new trees only in existing forest, preserving meadows.

SylvainDevidal committed:
Feature: New Trees Placement Algorithm "Forest"

LordAro commented:
Without looking at the actual code, a couple of things are immediately obvious:

* Commit message - https://wiki.openttd.org/Commit_style#Commit_message
* Coding style - https://wiki.openttd.org/Coding_Style
* Do not edit any language files other than english.txt, the others are only touched by the web translator (this should probably be in CONTRIBUTING, tbh)

See CONTRIBUTING.md for more details

LordAro commented:
Thanks very much! I'd recommend using rebase to squash everything together - the build system will have a go at building it then, once it's happy with the commit messages!

I'm not sure about that last commit, the one about docker complaining about lack of newline at EoL - the git diff looks much like you've removed them rather than added them. Regardless, try not to touch stuff that's unrelated to the PR

SylvainDevidal commented:
Hello LordAro,
Thank you for you quick comments.
I think I fixed most of the issues:
- Reverted changes to lang files except "english.txt"
- Fixed (and simplified) coding style
- Changed the commit message

But there is still an issue with docker.
It complains about new line at end of files that I have not changed (in fact, I changed them accidentely and reverted them).
I tried to fix the issue, but the message presists:

`[commit-checker] Running shell script
+ docker logs --follow 152ab8ac1e672452fdb2f1e999af1f32b71c52757e2278b776c3adb6076469aa
Validating commits
Branch is on top of master
*** b/projects/generate_vs141.vcxproj: No newline at end of file
*** b/projects/langs_vs141.vcxproj: No newline at end of file
*** b/projects/openttd_vs141.vcxproj: No newline at end of file
*** b/projects/settings_vs141.vcxproj: No newline at end of file
*** b/projects/settingsgen_vs141.vcxproj: No newline at end of file
*** b/projects/strgen_vs141.vcxproj: No newline at end of file
+ docker wait 152ab8ac1e672452fdb2f1e999af1f32b71c52757e2278b776c3adb6076469aa
+ exit 1
script returned exit code 1`

LordAro commented:
Weird. Regardless, if you rebase your PR (something like `git rebase -i origin/master`) and squash all the commits together, that should remove any changes to the project files, meaning that the checker won't complain

SylvainDevidal commented:
Well, I really Don't understand how to do.
When I rebase + squash all my commits, I get a nice log in my local copy with only the desired files updated.
But when I try to push it, it claims I need to pull before, and then I get some spaghetti… And still the error :(

LordAro commented:
Yeah, you're rewriting the commit history, so you'll need to "force-push" when pushing to a remote - just add -f to the command

SylvainDevidal commented:
Oh, ok.
Thank you :)
It looks like to work better now!

SylvainDevidal committed:
Merge pull request #1 from OpenTTD/master

Update: Translations from eints

michicc commented on src/tree_cmd.cpp:
```diff
@@ -305,6 +333,7 @@ void GenerateTrees()
 
 	switch (_settings_game.game_creation.tree_placer) {
 		case TP_ORIGINAL: i = _settings_game.game_creation.landscape == LT_ARCTIC ? 15 : 6; break;
+		case TP_FOREST: /* FALL THROUGH */
```
We have a proper macro for this now (`FALLTHROUGH`), to take advantage of some C++17 stuff if supported.

michicc commented:
In general, you shouldn't have merges in PRs. We will always squash or rebase a PR anyway to keep a linear history.

A merge only complicates reviewing as it can make the changes less clear and harder to check if the resulting rebase/squash will result in a working state. GitHub checks for conflicts, but this is is invariably only a textual check and not a logical one.

SylvainDevidal committed:
Codechange: Use of proper FALLTHROUGH macro

SylvainDevidal commented:
Hello michicc,

I Added the FALLTHROUGH macro instead of the comment.
The "coding style" section should be updated in Wiki then.

andythenorth commented:
<img width="1265" alt="forest_patch_2" src="https://user-images.githubusercontent.com/1780327/42964307-863b0242-8b8e-11e8-8a6d-6fa39141fc09.png">
<img width="777" alt="forest_patch_1" src="https://user-images.githubusercontent.com/1780327/42964309-88d44cac-8b8e-11e8-9d60-b0ade6d8d6dc.png">

nielsmh commented:
Those screenshots makes it look a bit strange, forests forming in perfect rectangles. The rest of the land also looks unnecessarily barren.

Can I suggest adding some more fuzziness so forest spread can be more uneven, and also have a minimal chance trees can spring up in the middle of nowhere to jumpstart new forests occasionally?

andythenorth commented:
Hi Sylvain!

Thanks for this, trees do need some love.

I'm not a core dev, but I tested your PR, and here are my observations :)

- as nielsm said, the forests make quite regular diamond shapes, with no trees at all on other tiles, this looks odd in the map / minimap, more randomness would be nice
- the world-gen menu items need some attention, 'none', 'original', 'improved', 'forest' is now quite confusing, hard for the player to understand which to pick

I hope you can keep contributing and the end result is nicer trees for the game!

SylvainDevidal committed:
Codechange: Merged "Improved" and "Forest" options.
Codechange: Added some alone trees during generation.
Codechange: Added some randomness to the tree placement distance around forests.

SylvainDevidal commented:
Hello andythenorth and nielsmh,

Thank you for your feedback.

I agree "Improved" and "Forest" are quite confusing. So I updated the "Improved" feature instead of creating a new one. From my point of view, the initial "Improved" and "Original" features were not making big difference as the trees were planted on all tiles of the map after a few décades.

So I changed this:
- Improved now plant fewer trees alone during map generation (16 times less)
- Improved now doesn't make trees spreading if there are not other near by trees

As a result, you can see now a few trees appearing in meadows, but they won't spread to cover the whole map.

For the diamond aspect of forests, if you look closely to the original "improved" forests, you'll see they were already diamonds. It was less ugly as there were also many random trees everywhere, so the forests limits were not much visible.

The problem is located in the function PlaceTreeGroups that I didn't changed. It chooses a tile, then plant many trees arround, but there is no "shape", so the result produces a square.
I currently don't have idea how to improve it.

SylvainDevidal commented:
I added some randomness to the forest tree placement. It looks like to produce something better.

TrueBrain commented on src/table/settings.ini:
```diff
@@ -2216,7 +2216,7 @@ from     = 30
 def      = 2
 min      = 0
-max      = 2
+max      = 3
```
Think this is a remainder of a past version now? :)

TrueBrain commented on src/tree_cmd.cpp:
```diff
@@ -193,11 +220,16 @@ static void PlaceTreeGroups(uint num_groups)
 			uint dist = abs(x) + abs(y);
+			uint max_dist = (_settings_newgame.game_creation.tree_placer == TP_IMPROVED) ? GB(r, 0, 3) + 8 : MAX_DISTANCE_FROM_CENTER;
+
+			if (dist > max_dist)
```
Please either put this on a single line, or add {} around it. This form only leads to errors over time :)

SylvainDevidal replied:
I splitted this line, now it's more clear

michicc replied:
I think TrueBrain meant the coding style for the `if`. Either make it a single line without any linebreaks, or use `{}`.

TrueBrain reviewed (dismissed):
Do you happen to have a screenshot how the latest result looks? :D

SylvainDevidal committed:
Fix: Revert max value for tree placement option

SylvainDevidal committed:
Fix: More random for tree placement in forest

SylvainDevidal commented:
Here is a new commit with a new picture :

![tree1](https://user-images.githubusercontent.com/6975555/43011700-8833f714-8c44-11e8-97ec-b114298aa613.png)

The "diamond" layout is attenuated

SylvainDevidal committed:
Codechange: code more readable

andythenorth commented:
Looks better to me.  What's it like in sub-tropic and arctic climates?  Sub-tropic has extra complication, no trees in desert :)

TrueBrain commented on src/tree_cmd.cpp:
```diff
@@ -193,11 +220,20 @@ static void PlaceTreeGroups(uint num_groups)
+			uint max_dist = MAX_DISTANCE_FROM_CENTER;
+
+			if (_settings_newgame.game_creation.tree_placer == TP_IMPROVED)
+				max_dist = GB(r, 16, 3) + 8;
```
Exactly what @michicc said; coding style, please never put a if statement with a single line body over two lines without {}. So either:

```if (_settings_newgame.game_creation.tree_placer == TP_IMPROVED) max_dist = GB(r, 16, 3) + 8;```

or

```
if (_settings_newgame.game_creation.tree_placer == TP_IMPROVED) {
    max_dist = GB(r, 16, 3) + 8;
}
```

Same goes for the line below :)
This to prevent future errors, where someone mistakes the indentation for a start of a new scope.

SylvainDevidal replied:
Oh, ok, got it!

SylvainDevidal committed:
Cleanup: Coding style fix

andythenorth commented:
Thx ;) 

I tested this some more.  I think the game is now not replacing trees that have died.  I tested this by running the game on fast forward for a while.  On my test map, most of the trees were gone (and not replaced) after about 50 years, with no replacements.  I didn't screenshot before/after, but I can run this again if that helps.

SylvainDevidal commented:
@andythenorth is there a way to "really fast forward" a game ?

When running in fast mode, one year still takes several minutes so it's very hard to check the effects of the patch on the long term

nielsmh commented:
Integrate the frame rate window changes that were merged a few days ago and check why it's running slow :)
If you zoom all the way in, make the game window small, and place the view in a corner of the map, you usually waste the least time on graphics rendering, which gets you the fastest passage of time in fast-forward.

andythenorth commented:
As nielsm mentioned, I place the view in a corner of the map.  

I also disable 'full animation' as it slows fast-forward on some platforms.  I also close all game windows (minimap etc).

It took around 10 minutes to run 50 years for me.

SylvainDevidal committed:
Fix: "Alone" trees are now replaced after death. They also can spread on their own tile.

SylvainDevidal commented:
Here is a quick fix for the "alone" tree disapearing.
1/ Now they can spread on their own tile : we get small grove that is much realistic (IMHO)
2/ Thus when they get old, they are replaced with a new one

andythenorth commented:
I've tested this again.  

- the changes from 23 July have improved the disappearing trees issue
- testing on FFWD, I'm not convinced that the issue is completely mitigated: 'forests' start out quite dense and uniform, then seem to thin out as time progresses in the game (see screenshots, 100 years of game time)

Is this better than trunk 'improved' algorithm?

- trunk 'improved' spams far too many trees on the map IMHO, it's overwhelming (I always toggle trees off in transparency for this reason)
- but the trees in this patch seem a bit too sparse IMHO

<img width="1222" alt="6848-jan-1900" src="https://user-images.githubusercontent.com/1780327/45234754-8f2e8d80-b2ce-11e8-8ddc-62c2b9ae7560.png">
<img width="1092" alt="6848-jan-2000" src="https://user-images.githubusercontent.com/1780327/45234755-8fc72400-b2ce-11e8-972f-922a8ead7b94.png">

TrueBrain commented:
We recently switched from Jenkins as CI to Azure Pipelines as CI. This means you need to rebase before this Pull Request will pass its checks. Sorry for the troubles!

andythenorth commented:
Thanks for this.  There's been no activity on this for some time, and as it stands, it doesn't look likely that it will go any further.  I'm closing it as we try to keep the open PR count low for OpenTTD, it helps us focus on things that are important and fun.  Feel free to discuss in irc or request re-opening if you disagree.  Thanks for contributing!
